rfd/0078-login-rules.md

```diff
@@ -0,0 +1,405 @@
+which may be better names (such as `if`) must unfortunately be avoided.
+
+#### `set(...items)`
+
+`set()` returns a set including all its arguments.
+Duplicates are not possible and will be filtered out on creation of the set if
+they are passed in.
+All items must be strings.
+As a special case, the empty string will not be included in the set.
+
+#### `set.contains(value)`
+
+`set.contains(value)` returns `true` if any item in the set is an exact match
+for `value`, else it returns `false`.
+`set("a", "b").contains("b")` returns `true`.
+
+#### `set.add(value)`
+
+`set.add(value)` returns a copy of the set with `value` added.
+`set("a", "b").add("c")` returns `("a", "b", "c")`.
+
+#### `set.remove(value)`
+
+`set.add(value)` returns a copy of the set with `value` removed.
+`set("a", "b", "c").remove("c")` returns `("a", "b")`.
+
+#### `dict(...pairs)`
+
+`dict()` returns a dictionary with keys of type `string` and values of type `set`.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+)
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("asparagus", "brocolli"),
+}
+```
+
+Arguments must be pairs, the first element will be the key,
+and the second element will be the value.
+
+#### `dict.add_values(key, ...values)`
+
+`dict.add_values(key, ...values)` returns a copy of the dict with the given values
+added to the set at `dict[key]`.
+If `dict[key]` is empty or it does not exist, it will be added with `values` as
+its only elements.
+
+```
+dict(
+  pair("fruits", set("apple")),
+).add("fruits", "banana").add("vegetables", "asparagus", "brocolli")
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("asparagus", "brocolli"),
+}
+```
+
+#### `dict.remove_keys(...keys)`
+
+`dict.remove_keys(...keys)` returns a copy of the dict with the given keys
+removed.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+).remove_keys("vegetables")
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+}
+```
+
+#### `dict.overwrite(key, values)`
+
+`dict.overwrite(key, values)` returns a copy of the dict with the given `key` set
+to `values`.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+).overwrite("vegetables", set("carrot")).overwrite("trees", set("aspen"))
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("carrot"),
+  "trees": ("aspen"),
+}
+```
+
+#### `ifelse(cond, value_if_true, value_if_false)`
+
+`ifelse(cond, value_if_true, value_if_false)` returns `value_if_true` if `cond`
+evaluates to `true`, else it returns `value_if_false`.
+`ifelse(set("a").contains("a"), set("b", "c"), set())`
+returns
+`("b", "c")`.
+
+#### `match(value, ...options)`
+
+`match(value, ...options)` returns the value of the first `option` which is a
```

Comment:
I have implemented it slightly differently, it`s:

`select` with no argument, it simply alternates between options:

```
select {
   case 
}
```

Reply:
@klizhentas I think we can't use `select` because it's a keyword and is a statement rather than an expression in Go, the parser really doesn't like it. For now I've tried to recreate this style but named it `choose`, let me know if you think there's a better way to name it